src/main/java/com/google/api/codegen/transformer/csharp/CSharpSurfaceNamer.java

```diff
@@ -446,4 +450,73 @@ public String getGrpcStreamingApiReturnTypeName(Method method, ModelTypeTable ty
     }
     throw new IllegalStateException("Invalid Synchronicity: " + synchronicity);
   }
+
+  @Override
+  public String getResourceOneofCreateMethod(ModelTypeTable typeTable, FieldConfig fieldConfig) {
+    String result = super.getResourceOneofCreateMethod(typeTable, fieldConfig);
+    return result.replaceFirst("IEnumerable<(\\w*)>(\\..*)", "$1$2");
+  }
+
+  @Override
+  public String makePrimitiveTypeNullable(String typeName, TypeRef type) {
+    return isPrimitive(type) ? typeName + "?" : typeName;
+  }
+
+  @Override
+  public boolean isPrimitive(TypeRef type) {
+    if (type.isRepeated()) {
+      return false;
+    }
+    switch (type.getKind()) {
+      case TYPE_BOOL:
+      case TYPE_DOUBLE:
+      case TYPE_ENUM:
+      case TYPE_FIXED32:
+      case TYPE_FIXED64:
+      case TYPE_FLOAT:
+      case TYPE_INT32:
+      case TYPE_INT64:
+      case TYPE_SFIXED32:
+      case TYPE_SFIXED64:
+      case TYPE_SINT32:
+      case TYPE_SINT64:
+      case TYPE_UINT32:
+      case TYPE_UINT64:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  @Override
+  public String getOptionalFieldDefaultValue(
+      FieldConfig fieldConfig, MethodTransformerContext context) {
+    // Need to provide defaults for primitives, enums, strings, and repeated (including maps)
+    TypeRef type = fieldConfig.getField().getType();
+    if (context.getFeatureConfig().useResourceNameFormatOption(fieldConfig)) {
+      if (type.isRepeated()) {
+        TypeName elementTypeName =
+            new TypeName(
+                getResourceTypeNameObject(fieldConfig.getResourceNameConfig()).toUpperCamel());
+        TypeNameConverter typeNameConverter = getTypeNameConverter();
+        TypeName enumerableTypeName = typeNameConverter.getTypeName("System.Linq.Enumerable");
+        TypeName emptyTypeName =
+            new TypeName(
+                enumerableTypeName.getFullName(),
+                enumerableTypeName.getNickname(),
+                "%s.Empty<%i>",
+                elementTypeName);
+        return TypedValue.create(emptyTypeName, "%s()")
+            .getValueAndSaveTypeNicknameIn((CSharpTypeTable) typeNameConverter);
+      } else {
+        return null;
+      }
+    } else {
+      if (type.isPrimitive() || type.isEnum() || type.isRepeated()) {
+        return context.getTypeTable().getInternalZeroValueAndSaveNicknameFor(type);
```

Comment:
As for the rationale of `getParamZeroValue`: without looking deeper, I was assuming this would appear in the method signature. I have since discovered that it only appears in the implementation.